### PR TITLE
feat(ticket #162) Added Mapper For Member List

### PR DESCRIPTION
**Pull request by muhammad-bilal-ali-khokhar**
### Description
This PR addressing a update in members table , where formmating `phone number` , user role  and added some formatters for feather uses.

### Implementation
- memberListMapper.ts
- memberRoleFormatter.ts
- phoneNumberFormatter
- stringFormater

### Screenshot
image

### CODE REVIEW

File: src/utils/stringFormater.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export const snakeToCapitalization = (item: string) => {
+  return item
+    .split("_")
+    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+    .join(" ");
+};
+
+export const capitalization = (item: string) => {
```
Comment (nadirhussain786) [inline src/utils/stringFormater.ts:8]: @muhammad-bilal-ali-khokhar  why we can’t fix this with css?

File: src/utils/stringFormater.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export const snakeToCapitalization = (item: string) => {
```
Comment (nadirhussain786) [inline src/utils/stringFormater.ts:1]: @muhammad-bilal-ali-khokhar  why we need this?

File: src/common/StatusBadge/UserStatusBadge.tsx [hunk 3]
```diff
@@ -24,7 +25,7 @@ export const UserStatusBadge = ({ status }: { status: UserStatus }) => {
     <Badge
       className={`capitalize flex justify-center ${background} ${text} hover:${background} hover:${text} hover:cursor-pointer`}
     >
-      {status}
+      {capitalization(status)}
```
Comment (nadirhussain786) [inline src/common/StatusBadge/UserStatusBadge.tsx:28]: @muhammad-bilal-ali-khokhar  Why do we need to use this `capitalization` function when it's already handled with CSS?

File: src/types/index.tsx [hunk 4]
```diff
@@ -1,11 +1,12 @@
...
+export type UserRole = "incharge" | "member" | "shift_incharge";
 export type Shifts = "A" | "B" | "C" | "D";
 export type UserStatus = "active" | "inactive" | "pending";
+
 export type Member = {
   shift: string;
   id: number;
   name: string;
-  phone: string;
+  phoneNumber: string;
```
Comment (nadirhussain786) [inline src/types/index.tsx:9]: @muhammad-bilal-ali-khokhar We should also consider the potential side effects of changing key entries.

**Review by nadirhussain786 — CHANGES_REQUESTED** [review]

File: src/utils/stringFormater.ts (hunk 1 above)
Comment (Monfernape) [reply src/utils/stringFormater.ts:8]: Good call Nadir.

File: src/types/index.tsx (hunk 4 above)
Comment (Monfernape) [reply src/types/index.tsx:9]: This looks fine. We did switch to `phoneNumber`, right? Good reminder though.

File: src/app/members/actions/createUser.ts [hunk 5]
```diff
@@ -20,7 +20,7 @@ export const createUser = async (values: UserBuilder) => {
 
   const invite_link = `${protocol}://${host}?token=${randomInvitedId}`;
 
-  const userPhoneNumber = formatPhoneNumber(values.phoneNumber);
+  const userPhoneNumber = intlNumberFormat(values.phoneNumber);
```
Comment (Monfernape) [inline src/app/members/actions/createUser.ts:23]: should we just rename this variable to `formattedPhoneNumber`?

File: src/types/index.tsx [hunk 6]
```diff
@@ -1,11 +1,12 @@
-export type UserRole = "incharge" | "member" | "shift-incharge";
+export type UserRole = "incharge" | "member" | "shift_incharge";
```
Comment (Monfernape) [inline src/types/index.tsx:1]: This is a big change, has this been tested?

File: src/utils/memberListMapper.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { Member } from "@/types";
+import { localNumberFormat } from "./phoneNumberFormatter";
+import { roleFormmater } from "./memberRoleFormatter";
+
+const MEMBER_LIST_MAPPER = (member: Member) => ({
```
Comment (Monfernape) [inline src/utils/memberListMapper.ts:5]: Functions can't be named in this manner. Why all uppercase?

File: src/utils/memberListMapper.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  ...member,
+  phoneNumber: localNumberFormat(member.phoneNumber),
+  role: roleFormmater(member.role),
+});
+
+export const memberListMapper = (members: Member[]) => {
+  const MemberList: Member[] = [];
+  members.forEach((member) => {
+    const mappedMember = MEMBER_LIST_MAPPER(member) as Member;
```
Comment (Monfernape) [inline src/utils/memberListMapper.ts:14]: Let's not use type assertion.

File: src/utils/memberListMapper.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import { Member } from "@/types";
```
Comment (Monfernape) [inline src/utils/memberListMapper.ts:1]: We do not need this file at all. If this isn't reusable, it doesn't make sense to keep it in separate file as utility.

**Review by Monfernape — CHANGES_REQUESTED** [review]

File: src/utils/stringFormater.ts (hunk 2 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/utils/stringFormater.ts:1]: we have some values in snakecase like `shift_incharge` , i want show it in table `Shift Incharge`

File: src/utils/stringFormater.ts (hunk 1 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/utils/stringFormater.ts:8]: css capitalize only changes the visual appearance, not the actual string value.when the capitalized value is needed programmatically, a function is required , in some case i need to convert values into function thats why i have created this function

File: src/app/members/actions/createUser.ts [hunk 10]
```diff
@@ -20,7 +21,7 @@ export const createUser = async (values: UserBuilder) => {
 
   const invite_link = `${protocol}://${host}?token=${randomInvitedId}`;
 
-  const userPhoneNumber = formatPhoneNumber(values.phoneNumber);
+  const userPhoneNumber = formattedPhoneNumber(values.phoneNumber);
```
Comment (Monfernape) [inline src/app/members/actions/createUser.ts:24]: LOL, you mistook my previous comment. What I meant was to update the variable name to `formattedPhoneNumber`. Function name `formatPhoneNumber` was fine.

File: src/app/members/components/MemberList.tsx [hunk 11]
```diff
@@ -24,16 +24,21 @@ import { Member } from "@/types";
...
 interface Props {
   data: Member[];
   success: boolean;
   message: string;
 }
+
 export const MemberList = ({ members }: { members: Props }) => {
   const { toast } = useToast();
-  const { data, success } = members;
+  const { data : membersList, success } = members;
```
Comment (Monfernape) [inline src/app/members/components/MemberList.tsx:38]: I'm not a fan of any variable name with list. Also, why are you getting `success` and `message` from props? I thought we removed that a long time ago. @nadirhussain786

File: src/app/members/components/MemberList.tsx [hunk 12]
```diff
@@ -58,7 +63,7 @@ export const MemberList = ({ members }: { members: Props }) => {
       accessorKey: "role",
       header: "Role",
       cell: ({ row }) => (
-        <div className="capitalize">{row.getValue("role")}</div>
+        <div className="capitalize">{formattedRole(row.getValue("role"))}</div>
```
Comment (Monfernape) [inline src/app/members/components/MemberList.tsx:66]: This becomes `formatRole`

File: src/types/index.tsx [hunk 13]
```diff
@@ -1,20 +1,21 @@
-export type UserRole = "incharge" | "member" | "shift-incharge";
+export type UserRoleType = "incharge" | "member" | "shift_incharge";
```
Comment (Monfernape) [inline src/types/index.tsx:1]: Why both enum and type?

File: src/types/index.tsx [hunk 14]
```diff
@@ -36,7 +37,7 @@ export type UserDetails = {
   cnic: string;
   ehad_duration: string;
   shift: string;
-  role: UserRole;
+  role: UserRoleType;
```
Comment (Monfernape) [inline src/types/index.tsx:40]: This looks wrong. Why switch to type instead of enum?

File: src/utils/stringFormater.ts (hunk 2 above)
Comment (Monfernape) [inline src/utils/stringFormater.ts:1]: Let's rename file to string. It's already in utils folder.

**Review by Monfernape — APPROVED** [review]
Approving to unblock but needs more work.

File: src/app/members/components/MemberList.tsx (hunk 11 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/app/members/components/MemberList.tsx:38]: Sir I have fixed , @nadirhussain786 i have assigned you a ticket for this [TICKET](https://github.com/Monfernape/shoes-rack/issues/217#issue-2698376510)

File: src/app/members/components/MemberList.tsx (hunk 12 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/app/members/components/MemberList.tsx:66]: fixed

File: src/types/index.tsx (hunk 13 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/types/index.tsx:1]: Sir , its just for readability purpose ,
we have two seprate ,
- define all types just in `type` file
- define all enums just in `constants` file 
- `UserRoleType` used only as type
- `UserRole` used only as value

File: src/types/index.tsx (hunk 14 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/types/index.tsx:40]: Sir , I didn't switch to type instead of enum , i just renamed type

File: src/utils/stringFormater.ts (hunk 2 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/utils/stringFormater.ts:1]: Sir these are common function so i have moved this to `Utils` file

File: src/utils/memberListMapper.ts (hunk 7 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/utils/memberListMapper.ts]: removed

**Review by nadirhussain786 — APPROVED** [review]

**Pull request merged**